File: src/AudioProcessor.ts
import { Effect, Either, Option, Ref, Schedule, Stream, Schema } from "effect";
import {
  AudioSource,
  BYTES_PER_SECOND,
  type AudioSourceId,
} from "./AudioSource.js";
import { OpenAIRealtime } from "./OpenAIRealtime.js";

const TARGET_BYTES = 15 * BYTES_PER_SECOND;
const COMMIT_BYTES = 3 * BYTES_PER_SECOND;

export class NoSourceError extends Schema.TaggedError<NoSourceError>()(
  "NoSourceError",
  {}
) {}

class SourceClearedError extends Schema.TaggedError<SourceClearedError>()(
  "SourceClearedError",
  {}
) {}

export class AudioProcessor extends Effect.Service<AudioProcessor>()(
  "AudioProcessor",
  {
    accessors: true,
    effect: Effect.gen(function* () {
      yield* Effect.log("Funny Radio - Sarcastic News Transformer");

      const waitForSource = AudioSource.currentSource.pipe(
        Effect.flatMap(Either.fromOption(() => new NoSourceError())),
        Effect.retry(Schedule.spaced("1 second"))
      );

      const processAudio = (sourceId: AudioSourceId) =>
        Effect.gen(function* () {
          yield* Effect.log(
            `Source selected: ${sourceId}, starting processing...`
          );

          const openai = yield* OpenAIRealtime;
          const accumulatedBytes = yield* Ref.make(0);
          const bytesSinceLastCommit = yield* Ref.make(0);

          const audioStream = yield* AudioSource.getStream();
          yield* audioStream.pipe(
            Stream.runForEach((chunk) =>
              Effect.gen(function* () {
                // Check if source was cleared or changed
                yield* AudioSource.currentSource.pipe(
                  Effect.flatMap(
                    Either.fromOption(() => new SourceClearedError())
                  ),
                  Effect.filterOrFail(
                    (current) => current === sourceId,
                    () => new SourceClearedError()
                  )
                );

                yield* openai.appendAudio(chunk.toString("base64"));

                const acc = yield* Ref.updateAndGet(
                  accumulatedBytes,
                  (n) => n + chunk.length
                );
                const since = yield* Ref.updateAndGet(
                  bytesSinceLastCommit,
                  (n) => n + chunk.length
                );

                if (since >= COMMIT_BYTES && acc < TARGET_BYTES) {
                  yield* openai.commitBuffer();
                  yield* Ref.set(bytesSinceLastCommit, 0);
                }

                if (acc >= TARGET_BYTES) {
                  yield* Effect.log(
                    `Requesting response (${(acc / BYTES_PER_SECOND).toFixed(1)}s of audio)`
                  );
                  yield* openai.commitBuffer();
                  yield* openai.requestResponse();
                  yield* Ref.set(accumulatedBytes, 0);
                  yield* Ref.set(bytesSinceLastCommit, 0);
                }
              })
            )
          );
        }).pipe(
          Effect.catchTag("SourceClearedError", () =>
            Effect.log("Source cleared, stopping audio processing")
          )
        );

      const run = Effect.gen(function* () {
        yield* Effect.log(
          "Audio processor initialized, waiting for source selection..."
        );

        // Loop: wait for source, process, repeat when cleared
        return yield* waitForSource.pipe(
          Effect.flatMap(processAudio),
          Effect.catchAllCause((cause) =>
            Effect.logError("Audio processing failed, restarting...", cause)
          ),
          Effect.repeat(Schedule.spaced("1 second"))
        );
      });

      return { run } as const;
    }),
  }
) {}

File: go/audio_processor.go
package main

import (
	"context"
	"encoding/base64"
	"fmt"
	"log"
	"time"
)

const (
	TargetBytes = 15 * BytesPerSecond // 15 seconds of audio before requesting response
	CommitBytes = 3 * BytesPerSecond  // 3 seconds between buffer commits
)

// AudioProcessor reads audio from the selected source and streams it to
// OpenAI via the Realtime API.
type AudioProcessor struct {
	source *AudioSource
	openai *OpenAIRealtime
}

func NewAudioProcessor(source *AudioSource, openai *OpenAIRealtime) *AudioProcessor {
	return &AudioProcessor{source: source, openai: openai}
}

// Run is the main processing loop. It waits for a source to be selected,
// processes audio, and restarts on error or source change.
func (ap *AudioProcessor) Run(ctx context.Context) {
	log.Println("Audio processor initialized, waiting for source selection...")

	for {
		select {
		case <-ctx.Done():
			return
		default:
		}

		// Wait for a source to be selected
		sourceID := ap.waitForSource(ctx)
		if sourceID == nil {
			return // context cancelled
		}

		// Process audio for this source
		err := ap.processAudio(ctx, *sourceID)
		if err != nil {
			log.Printf("Audio processing failed, restarting... error: %v", err)
		}

		// Brief pause before retrying
		select {
		case <-time.After(time.Second):
		case <-ctx.Done():
			return
		}
	}
}

func (ap *AudioProcessor) waitForSource(ctx context.Context) *AudioSourceID {
	for {
		if id := ap.source.CurrentSource(); id != nil {
			return id
		}
		select {
		case <-time.After(time.Second):
		case <-ctx.Done():
			return nil
		}
	}
}

func (ap *AudioProcessor) processAudio(ctx context.Context, sourceID AudioSourceID) error {
	log.Printf("Source selected: %s, starting processing...", sourceID)

	// Create a cancellable context for this processing session
	procCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	audioCh, err := ap.source.Stream(procCtx)
	if err != nil {
		return fmt.Errorf("failed to start audio stream: %w", err)
	}

	accumulated := 0
	sinceCommit := 0
	chunkCount := 0
	throughputStart := time.Now()

	for chunk := range audioCh {
		// Check that the source hasn't changed
		current := ap.source.CurrentSource()
		if current == nil || *current != sourceID {
			log.Println("Source cleared, stopping audio processing")
			return nil
		}

		// Encode and send to OpenAI
		b64 := base64.StdEncoding.EncodeToString(chunk)
		if err := ap.openai.AppendAudio(b64); err != nil {
			return fmt.Errorf("failed to append audio: %w", err)
		}

		accumulated += len(chunk)
		sinceCommit += len(chunk)
		chunkCount++

		// Periodic commit (every 3 seconds of audio)
		if sinceCommit >= CommitBytes && accumulated < TargetBytes {
			if err := ap.openai.CommitBuffer(); err != nil {
				return fmt.Errorf("failed to commit buffer: %w", err)
			}
			sinceCommit = 0
		}

		// Request response after 15 seconds of audio
		if accumulated >= TargetBytes {
			elapsed := time.Since(throughputStart)
			bytesTotal := accumulated
			log.Printf("Requesting response (%.1fs of audio)", float64(accumulated)/float64(BytesPerSecond))
			if elapsed > 0 {
				chunksPerSec := float64(chunkCount) / elapsed.Seconds()
				bytesPerSec := float64(bytesTotal) / elapsed.Seconds()
				log.Printf("[KPI] chunk_throughput: %.1f chunks/s, %.0f bytes/s (%.1fx realtime)",
					chunksPerSec, bytesPerSec, bytesPerSec/float64(BytesPerSecond))
			}
			if err := ap.openai.CommitBuffer(); err != nil {
				return fmt.Errorf("failed to commit buffer: %w", err)
			}
			if err := ap.openai.RequestResponse(); err != nil {
				return fmt.Errorf("failed to request response: %w", err)
			}
			accumulated = 0
			sinceCommit = 0
			chunkCount = 0
			throughputStart = time.Now()
		}
	}

	return nil
}
